### ISSUE TITLE
No way to detect shutdown on Azure App Service on Linux

### ISSUE THREAD

**Issue by tuukkapuranen** [dotnet/aspnetcore#29923]
### Describe the bug

When running Azure Web App using App Service on Linux, the `IHostApplicationLifeTime` `ApplicationStopping` event is not fired.

### To Reproduce

This is the same issue as https://github.com/dotnet/aspnetcore/issues/18259, and [here are the steps needed to reproduce the behavior](https://github.com/dotnet/aspnetcore/issues/18259#issuecomment-577648350).

### Further technical details

To add further details to the previous report, we observed that this behavior is present when running on .NET Core 3.1 or .NET 5.0.

The behavior is also different when running the Web App using App Service on Windows. In that case, the `ApplicationStopping` event is fired. The event is fired on service restarts and on configuration updates.

In addition, there seems to be no workaround as none of the following methods appear to detect shutdown on App Service on Linux:
 - [IHostApplicationLifetime.ApplicationStopping](https://docs.microsoft.com/en-us/dotnet/api/microsoft.extensions.hosting.ihostapplicationlifetime.applicationstopping?view=dotnet-plat-ext-5.0)
 - [IHostApplicationLifetime.ApplicationStopped](https://docs.microsoft.com/en-us/dotnet/api/microsoft.extensions.hosting.ihostapplicationlifetime.applicationstopped?view=dotnet-plat-ext-5.0)
 - [IHostedService.StopAsync](https://docs.microsoft.com/en-us/dotnet/api/microsoft.extensions.hosting.ihostedservice.stopasync?view=dotnet-plat-ext-5.0)
 - [AppDomain.ProcessExit](https://docs.microsoft.com/en-us/dotnet/api/system.appdomain.processexit?view=net-5.0) using `AppDomain.CurrentDomain.ProcessExit`
 - [AssemblyLoadContext.Unloading](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.loader.assemblyloadcontext.unloading?view=net-5.0) using `AssemblyLoadContext.Default.Unloading`
 - [Console.CancelKeyPress](https://docs.microsoft.com/en-us/dotnet/api/system.console.cancelkeypress?view=net-5.0)

**Comment by Tratcher** [comment]
From what we've been able to determine, this is an issue in the App Service infrastructure. @bradygaster any progress on that end?

**Comment by jkotalik** [comment]
Duplicate of https://github.com/dotnet/aspnetcore/issues/22272

**Comment by tuukkapuranen** [comment]
I am unsure whether this issue is duplicate of https://github.com/dotnet/aspnetcore/issues/22272. The mentioned issue is about Azure WebJobs running on the App Service on Windows. This issue is describing different behavior. Namely, the Web App itself raising the lifetime events. For example, I can confirm that the StopAsync is called correctly on the Web App itself on Windows host with the following code:

```
internal class LifetimeEventsHostedService : IHostedService
{
    private readonly TelemetryClient _logger;

    public LifetimeEventsHostedService(IServiceProvider services)
    {
        _logger = services.GetService<TelemetryClient>();
    }

    public Task StartAsync(CancellationToken cancellationToken)
    {
        _logger.TrackEvent("StartAsync");
        return Task.CompletedTask;
    }

    public Task StopAsync(CancellationToken cancellationToken)
    {
        _logger.TrackEvent("StopAsync");
        _logger.Flush();
        Thread.Sleep(5000);
        return Task.CompletedTask;
    }
}
```

when registered as a hosted service with the following.

`services.AddHostedService<LifetimeEventsHostedService>();`

Note that these hosted services are not WebJobs.

The behavior described in this issue is that the above example - and indeed any of the mentioned lifetime events for detecting App Service shutdown - do not work when hosted on Linux. The mentioned duplicate issue concerns only hosting on Windows.

To summarize, this issue describes a difference in behavior of the Web Apps depending on the hosting OS, and has nothing to do with WebJobs. On Windows, all lifetime events are raised on the Web App, on Linux - the shutdown events are not.

**Comment by Tratcher** [comment]
There was also #24569, but we suspected the same root infrastructure issue in app service.
